### PR TITLE
planner: fix bug the inner child can't keep order in index join

### PR DESCRIPTION
**Pull request by lzmhhh123**
### What problem does this PR solve? 
As the title says.

### What is changed and how it works?
Add the `extraHandleCol` to cop task.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

Side effects

 - Possible performance regression
 - Increased code complexity

### CODE REVIEW

**Comment by zz-jason** [comment]
do we need to cherry pick this bug-fix to the release branches?

**Comment by lzmhhh123** [comment]
No, It's just a bug in the master branch.

File: planner/core/exhaust_physical_plans.go [hunk 1]
```diff
@@ -716,6 +716,21 @@ func (p *LogicalJoin) constructInnerIndexScanTask(
...
+			cop.extraHandleCol, cop.doubleReadNeedProj = ts.appendExtraHandleCol(ds)
+			schemaColumns := make([]*expression.Column, len(is.dataSourceSchema.Columns))
+			copy(schemaColumns, is.dataSourceSchema.Columns)
+			ts.schema = ts.schema.Clone()
+			ts.schema.Columns = schemaColumns
+			// If the double read need projection, the datasource schema will append a handle column during appendExtraHandleCol,
+			// So we need to delete the extra column in schema to avoid wrong plan.
+			if cop.doubleReadNeedProj {
+				defer func() {
+					ds.schema.Columns = ds.schema.Columns[:len(ds.schema.Columns)-1]
```
Comment (winoros) [inline planner/core/exhaust_physical_plans.go]: where did you change DataSource's schema?
The added projection won't affect your execution logic?

File: planner/core/exhaust_physical_plans.go [hunk 2]
```diff
@@ -716,6 +716,21 @@ func (p *LogicalJoin) constructInnerIndexScanTask(
 		// On this way, it's double read case.
 		ts := PhysicalTableScan{Columns: ds.Columns, Table: is.Table, TableAsName: ds.TableAsName}.Init(ds.ctx, ds.blockOffset)
 		ts.SetSchema(is.dataSourceSchema)
+		// If inner cop task need keep order, the extraHandleCol should be set.
+		if cop.keepOrder {
+			cop.extraHandleCol, cop.doubleReadNeedProj = ts.appendExtraHandleCol(ds)
+			schemaColumns := make([]*expression.Column, len(is.dataSourceSchema.Columns))
+			copy(schemaColumns, is.dataSourceSchema.Columns)
+			ts.schema = ts.schema.Clone()
```
Comment (winoros) [inline planner/core/exhaust_physical_plans.go]: Why you clone it and then reset it?
```
ts.schema = expression.NewSchema(schemaColumns)
```

File: planner/core/exhaust_physical_plans.go (hunk 1 above)
Comment (lzmhhh123) [reply planner/core/exhaust_physical_plans.go]: 1. DataSource's schema is changed in `ts.appendExtraHandleCol(ds)`
2. It does affect execution. So I will change the executing logic in PR of index merge join.

Comment (winoros) [reply planner/core/exhaust_physical_plans.go]:
```
// If there is a table reader which needs to keep order, we should append a pk to table scan.
func (ts *PhysicalTableScan) appendExtraHandleCol(ds *DataSource) {
	if len(ds.schema.TblID2Handle) > 0 {
		return
	}
	pkInfo := model.NewExtraHandleColInfo()
	ts.Columns = append(ts.Columns, pkInfo)
	handleCol := ds.newExtraHandleSchemaCol()
	ts.schema.Append(handleCol)
	ts.schema.TblID2Handle[ds.tableInfo.ID] = []*expression.Column{handleCol}
}
```

```
func (ds *DataSource) newExtraHandleSchemaCol() *expression.Column {
	return &expression.Column{
		DBName:   ds.DBName,
		TblName:  ds.tableInfo.Name,
		ColName:  model.ExtraHandleName,
		RetType:  types.NewFieldType(mysql.TypeLonglong),
		UniqueID: ds.ctx.GetSessionVars().AllocPlanColumnID(),
		ID:       model.ExtraHandleID,
	}
}
```

As you can see, datasource's schema is not changed. And it should not be changed.

Comment (lzmhhh123) [reply planner/core/exhaust_physical_plans.go]: `ts.schema.Append(handleCol)` this line will change.

Comment (lzmhhh123) [reply planner/core/exhaust_physical_plans.go]: `ts.schema` and `ds.schema` are same pointer here.

Comment (winoros) [reply planner/core/exhaust_physical_plans.go]:
```
// If there is a table reader which needs to keep order, we should append a pk to table scan.
func (ts *PhysicalTableScan) appendExtraHandleCol(ds *DataSource) {
	if len(ds.schema.TblID2Handle) > 0 {
		return
	}
	pkInfo := model.NewExtraHandleColInfo()
	ts.Columns = append(ts.Columns, pkInfo)
	handleCol := ds.newExtraHandleSchemaCol()
	ts.schema.Append(handleCol)
	ts.schema.TblID2Handle[ds.tableInfo.ID] = []*expression.Column{handleCol}
}
```

```
func (ds *DataSource) newExtraHandleSchemaCol() *expression.Column {
	return &expression.Column{
		DBName:   ds.DBName,
		TblName:  ds.tableInfo.Name,
		ColName:  model.ExtraHandleName,
		RetType:  types.NewFieldType(mysql.TypeLonglong),
		UniqueID: ds.ctx.GetSessionVars().AllocPlanColumnID(),
		ID:       model.ExtraHandleID,
	}
}
```

As you can see, datasource's schema is not changed. And it should not be changed.

Comment (winoros) [reply planner/core/exhaust_physical_plans.go]: ok then, you should first let ts.schema and ds.schema not be the same thing. Then don't change ds.schema.

Comment (lzmhhh123) [reply planner/core/exhaust_physical_plans.go]: Yeah. That's a better way.

**Pull request merged**